### ISSUE TITLE
Potential deadlock in RestierContainerBuilder.AddRestierService

### ISSUE THREAD

**Issue by jspuij** [jspuij/RESTier#28]
This method contains the following code:

``` 
IEdmModel modelFactory(IServiceProvider sp)
            {
                var api = sp.GetService<ApiBase>();
                var model = api.GetModelAsync(default).GetAwaiter().GetResult();
                return model;
            }
```

This delegate will block the current thread and potentially deadlock in asp.net. Full reason why:
https://blog.stephencleary.com/2012/07/dont-block-on-async-code.html

See also: https://github.com/jspuij/RESTier/issues/11

**Comment by robertmclaws** [comment]
I was led to believe that GetAwaiter().GetResult() is the safe way to ensure you can get the result without deadlocking your code. I could be totally wrong about this, I have no idea. But I needed to make GetModel async so it could access some async-only functions, and I could not rewire the stuff that calls GetModelAsync without making the whole platform async, so I was kinda stuck. If you have a better solution right here that doesn't require massive codebase changes, I'm all for it.

**Comment by jspuij** [comment]
Nope, the only reason why it is a little better, is that it does not return an Aggregate Exception but the individual exceptions.

Wrapping this in a Task.run is not really advisable as well, because the container is building, and I'm pretty sure that the DI implementation is not necessarily thread safe nor should it be a requirement. This is a risk anyway, because every Configurawait(false) during async code might possibly be scheduled on a threadpool thread. 

You're correct in that the best way to prevent this is to make the entire library async. But even then, container composition should be all about instantiation and wiring without I/O. So I'd rather deterministically build the model asynchronously before or after container is built, than trigger it during container composition.
